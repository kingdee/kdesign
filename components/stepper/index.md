---
category: Components
type: 表单
order: 13
title: Stepper
subtitle: 步进器
---

步进器是一种在指定的数值范围内增减步进值的组件。

## 使用场景

- 数值范围较大或无上下限制的情况下

- 需要强化用户键入数字时，可采用嵌入式步进器

## API

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| decimalLength | 小数位数 | number | `-` | `-` | 1.0.0 |
| digitLength | 数字的位数 | number | `-` | `-` | 1.0.0 |
| formatter | 指定输入框展示值的格式 | function(value: string): string | `-` | `-` | 1.0.0 |
| max | 最大值 | number | `-` | `-` | 1.0.0 |
| maxMark | 最大值比较符 | string | `]` | `)` \| `]` | 1.0.0 |
| min | 最小值 | number | `-` | `-` | 1.0.0 |
| minMark | 最小值比较符 | string | `(` | `(` \| `[` | 1.0.0 |
| mustInPrecisionScope | 输入限制在精度范围 | boolean | `true` | `true`\|`false` | 1.0.0 |
| mustInScope | 输入内容是否必须在设定的数字范围内(不在范围内则不允许输入) | boolean | `false` | `true`\|`false` | 1.0.0 |
| onChange | 输入框内容变化时的回调 | (event) => void | `-` | `-` | 1.0.0 |
| showDecimalTailZero | 是否显示小数尾部 0 | boolean | `false` | `true`\|`false` | 1.0.0 |
| step | 步长 | number | `1` | `-` | 1.0.0 |
| stepBtnClassName | 步进按钮样式 | string | `-` | `-` | 1.0.0 |
| symbol | 货币符号 | string | `-` | `-` | 1.0.0 |
| type | 步进器类型 | string | `base` | `embed` `base` | 1.0.0 |
| zeroShow | 为零是否显示 | boolean | `false` | `true` `false` | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-stepper-input-color | --kd-g-color-text-primary | #212121 |
|  | --kd-c-stepper-icon-color | --kd-g-color-text-secondary | #666 |
|  | --kd-c-stepper-icon-color-hover | --kd-g-color-theme | #5582f3 |
|  | --kd-c-stepper-color-border-strong | --kd-g-color-border-strong | #d9d9d9 |
|  | --kd-c-stepper-embed-icon-color-background | - | transparent |
|  | --kd-c-stepper-icon-color-disabled | --kd-g-color-disabled | #b2b2b2 |
| font | --kd-c-stepper-input-font-size | - | 12px |
| motion | --kd-c-stepper-motion-duration | --kd-g-duration | 0.3s |
| sizing | --kd-c-stepper-input-small-sizing-height | - | 20px |
|  | --kd-c-stepper-input-middle-sizing-height | - | 28px |
|  | --kd-c-stepper-input-large-sizing-height | - | 36px |
|  | --kd-c-stepper-padding-left | - | 9px |
|  | --kd-c-stepper-input-padding-left | - | 0px |
|  | --kd-c-stepper-sizing-height | - | 28px |
| spacing | --kd-c-stepper-input-spacing-padding-horizontal | - | 9px |
