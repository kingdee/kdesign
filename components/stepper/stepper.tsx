import React, { FunctionComponentElement, useContext, useRef, useCallback, useState } from 'react'
import classNames from 'classnames'
import Big from 'big.js'
import InputNumber, { InputNumberProps } from '../input-number'
import Icon from '../icon'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'
import { tuple } from '../_utils/type'
import { omit } from '../_utils/omit'
import { useMergedState } from '../_utils/hooks'
import { isExp } from '../_utils/numberUtil'

Big.PE = 40
Big.NE = -40

export const StepTypes = tuple('embed', 'base')
export type StepType = typeof StepTypes[number]
export interface StepperProps extends InputNumberProps {
  type?: StepType
  step?: number
  stepBtnClassName?: string
}

const InternalStepper = (props: StepperProps, ref: unknown): FunctionComponentElement<StepperProps> => {
  const { compDefaultProps: userDefaultProps, getPrefixCls, prefixCls } = useContext(ConfigContext)
  const inputNumberProps = getCompProps('Stepper', userDefaultProps, props)
  const {
    prefixCls: customPrefixcls,
    onChange,
    step,
    type,
    stepBtnClassName,
    max,
    min,
    disabled,
    value: propsValue,
    defaultValue,
  } = inputNumberProps
  const inputPrefixCls = getPrefixCls!(prefixCls, 'inputNumber', customPrefixcls)

  devWarning(typeof step !== 'number', 'stepper', `step必须为一个数值`)

  const thisInputNumberRef = useRef<any>()
  const stepperrref = useRef<any>()
  const inputNumberRef = (ref as any) || thisInputNumberRef
  const stepMouseDownDelayTimer = useRef<any>(null)
  const stepMouseDownIntervalTimer = useRef<any>(null)

  const [showMinusdisabled, setMinusdisabled] = useState(false)
  const [showPlusdisabled, setPlusdisabled] = useState(false)
  const [value, setValue] = useMergedState(undefined, {
    value: propsValue,
    defaultValue,
  })

  // class
  const getStepBtnClassNames = (stepType: StepType, btnType?: 'decrease' | 'increase', isDisabled = false): string => {
    return classNames({
      [`${inputPrefixCls}-${stepType}Step`]: true,
      [`${inputPrefixCls}-${stepType}Step-disabled`]: isDisabled,
      [`${inputPrefixCls}-${stepType}Step-${btnType}`]: btnType,
      [`${inputPrefixCls}-${stepType}Step-small`]: inputNumberProps.size === 'small',
      [`${inputPrefixCls}-${stepType}Step-middle`]: inputNumberProps.size === 'middle',
      [`${inputPrefixCls}-${stepType}Step-large`]: inputNumberProps.size === 'large',
      [stepBtnClassName]: stepType === 'base' && stepBtnClassName,
    })
  }

  const getStepPrefix = (): React.ReactNode => {
    if (type === 'embed') {
      return null
    }
    const decreaseClassName = getStepBtnClassNames('base', 'decrease', showPlusdisabled || disabled)

    return (
      <span
        className={decreaseClassName}
        onMouseDown={() => {
          if (disabled) return
          handleStepMouseDown('minus')
        }}
      >
        <Icon className={`${inputPrefixCls}-icon`} type="reduce" />
      </span>
    )
  }

  const getStepSuffix = (): React.ReactNode => {
    const increaseClassName = getStepBtnClassNames('base', 'increase', showMinusdisabled || disabled)
    return type === 'embed' ? (
      getEmbedStepView()
    ) : (
      <span
        className={increaseClassName}
        onMouseDown={() => {
          if (disabled) return
          handleStepMouseDown('plus')
        }}
      >
        <Icon className={`${inputPrefixCls}-icon`} type="add" />
      </span>
    )
  }

  const isLegal = (value: any) => {
    const reg = /^[0-9]*\.?[0-9]+$/
    return reg?.test(value)
  }

  type StepBtnType = 'plus' | 'minus'
  const handleStepChang = (type: StepBtnType) => {
    const stepNum = parseFloat(step)
    if (typeof stepNum !== 'number') {
      return false
    }
    const startingNumber = isExp(stepperrref.current.value)
      ? Big(stepperrref.current.value).valueOf()
      : stepperrref.current.value || inputNumberProps.min || 0
    const currentValue = new Big(startingNumber)[type](stepNum)
    const calculationResults = currentValue.valueOf()
    const legalNumber = stepperrref.current.verifiValue(calculationResults)
    if (legalNumber === false) {
      return false
    }
    const nextPlus = new Big(calculationResults).plus(stepNum)
    const nextMinus = new Big(calculationResults).minus(stepNum)
    if (isLegal(max)) {
      const maxBig = new Big(max)
      if (nextPlus.cmp(maxBig) === 1) {
        setMinusdisabled(true)
      } else {
        setMinusdisabled(false)
      }
      if (currentValue.cmp(maxBig) === 1) {
        return false
      }
    } else {
      setMinusdisabled(false)
    }
    if (isLegal(min)) {
      const minBig = new Big(min)
      if (minBig.cmp(nextMinus) === 1) {
        setPlusdisabled(true)
      } else {
        setPlusdisabled(false)
      }
      if (minBig.cmp(currentValue) === 1) {
        return false
      }
    } else {
      setPlusdisabled(false)
    }
    props?.value === undefined && setValue(legalNumber)
    onChange && onChange({ target: { value: legalNumber } })
  }

  const setpLongPress = (type: StepBtnType) => {
    stepMouseDownIntervalTimer.current = setInterval(() => {
      handleStepChang(type)
    }, 300)
  }

  const handleStepMouseDown = (type: StepBtnType) => {
    handleStepChang(type)
    stepMouseDownDelayTimer.current = setTimeout(() => {
      setpLongPress(type)
    }, 300)
    document.addEventListener('mouseup', clearAllTimer, true)
    return false
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const number = parseFloat(event.target.value)
    if (
      String(value).length &&
      event.target.value?.length >= String(value).length &&
      event.target.value !== '' &&
      props?.value !== undefined &&
      ((max && max < number) || (min && min > number))
    ) {
      return
    }
    props?.value === undefined && setValue(event.target.value)
    props?.value === undefined && stepperrref.current.setValue(event.target.value)
    onChange && onChange(event)
  }

  const clearAllTimer = useCallback(() => {
    stepMouseDownDelayTimer.current && clearTimeout(stepMouseDownDelayTimer.current)
    stepMouseDownIntervalTimer.current && clearInterval(stepMouseDownIntervalTimer.current)
    document.removeEventListener('mouseup', clearAllTimer)
    return false
  }, [stepMouseDownDelayTimer, stepMouseDownIntervalTimer])

  const getEmbedStepView = (): React.ReactNode => {
    const embedStepClassName = getStepBtnClassNames('embed')
    const plusClassName = classNames(`${inputPrefixCls}-embedStep-plus`, {
      [stepBtnClassName]: stepBtnClassName,
      [`${inputPrefixCls}-embedStep-disabled`]: showMinusdisabled || disabled,
    })
    const minusClassName = classNames(`${inputPrefixCls}-embedStep-minus`, {
      [stepBtnClassName]: stepBtnClassName,
      [`${inputPrefixCls}-embedStep-disabled`]: showPlusdisabled || disabled,
    })
    return (
      <span className={embedStepClassName}>
        <span
          className={plusClassName}
          onMouseDown={() => {
            if (disabled) return
            handleStepMouseDown('plus')
          }}
        >
          <Icon className={`${inputPrefixCls}-icon`} type="arrow-up" />
        </span>
        <span
          className={minusClassName}
          onMouseDown={() => {
            if (disabled) return
            handleStepMouseDown('minus')
          }}
        >
          <Icon className={`${inputPrefixCls}-icon`} type="arrow-down" />
        </span>
      </span>
    )
  }

  return (
    <InputNumber
      {...omit(inputNumberProps, ['step', 'type', 'stepBtnClassName', 'value', 'defaultValue'])}
      ref={inputNumberRef}
      className={`${inputPrefixCls}-stepper`}
      stepperrref={stepperrref}
      value={value}
      prefix={getStepPrefix()}
      suffix={getStepSuffix()}
      onChange={handleChange}
    />
  )
}
const Stepper = React.forwardRef<unknown, StepperProps>(InternalStepper)
Stepper.displayName = 'Stepper'
export default Stepper
