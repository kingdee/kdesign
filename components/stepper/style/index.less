@import '../../style/themes/index';
@import './mixin.less';

@inputNumber-prefix-cls: ~'@{kd-prefix}-inputNumber';
.@{inputNumber-prefix-cls} {
  width: 100px;
  height: @stepper-input-middle-sizing-height;

  input {
    text-align: center;
    color: @stepper-input-color;
    margin: 0 4px;
    font-size: @stepper-input-font-size;
  }

  &-icon {
    font-size: 14px;
  }

  &-baseStep {
    display: flex;
    justify-content: center;
    align-items: center;
    box-sizing: border-box;
    height: 100%;
    cursor: pointer;

    &-small {
      .base-step-size(@stepper-input-small-sizing-height);
    }

    &-middle {
      .base-step-size(@stepper-input-middle-sizing-height);
    }

    &-large {
      .base-step-size(@stepper-input-large-sizing-height);
    }

    .@{inputNumber-prefix-cls}-icon {
      color: @stepper-icon-color;
      transition: color @stepper-motion-duration;
      font-size: 16px;

      &:hover {
        color: @stepper-icon-color-hover;
      }
    }

    &:hover:not(&-disabled) .@{inputNumber-prefix-cls}-icon {
      color: @stepper-icon-color-hover;
    }

    &-disabled {
      cursor: not-allowed;
      color: @stepper-icon-color-disabled;

      .@{inputNumber-prefix-cls}-icon {
        color: unset !important;
      }
    }
  }

  &-baseStep-decrease {
    .step-style(right, left);
  }

  &-baseStep-increase {
    .step-style(left, right);
  }

  &-embedStep {
    display: flex;
    flex-direction: column;
    justify-content: center;
    height: 100%;
    border-radius: 0 1px 1px 0;
    background-color: @stepper-embed-icon-color-background;
    overflow: hidden;
    .step-style(left, right);
    border: none;

    span {
      height: 50%;
      display: flex;
      align-items: center;
      justify-content: center;
      border-left: 1px solid @stepper-color-border-strong;
      border-bottom: 1px solid @stepper-color-border-strong;
      cursor: pointer;
      font-size: 12px;

      .@{inputNumber-prefix-cls}-icon {
        transition: color @stepper-motion-duration;
      }

      &:hover .@{inputNumber-prefix-cls}-icon {
        color: @stepper-icon-color-hover;
      }

      &:last-child {
        border-bottom: none;
      }
    }

    &-disabled {
      cursor: not-allowed !important;
      color: @stepper-icon-color-disabled;

      .@{inputNumber-prefix-cls}-icon {
        color: unset !important;
      }
    }
  }
}
