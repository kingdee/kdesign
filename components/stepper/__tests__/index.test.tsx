import React from 'react'
import { render, mount } from 'enzyme'
import Stepper from '../index'
import { StepTypes } from '../stepper'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'

describe('Stepper', () => {
  // 1. mount test
  describe('1. mount test', () => {
    mountTest(Stepper)
    StepTypes.forEach((type) => {
      mountTest(() => <Stepper type={type} />)
    })
  })

  // 2.render test
  describe('2. render test', () => {
    it('renders correctly', () => {
      expect(render(<Stepper />)).toMatchSnapshot()
      StepTypes.forEach((type) => {
        expect(render(<Stepper type={type} />)).toMatchSnapshot()
      })
    })
  })

  // 3. warns in component

  // 4. render null or undefined without errors
  // describe('4. render null or undefined without errors', () => {
  //   it('render null or undefined without errors', () => {
  //     expect(
  //       mount(
  //         <Stepper type="base">
  //           {null}
  //           {undefined}
  //         </Stepper>,
  //       ),
  //     ).toMatchSnapshot()
  //   })
  // })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      expect(Stepper.displayName).toBe('Stepper')
    })
  })

  // 6. api test
  describe('6. api test', () => {
    it('should class use right', () => {
      const BaseStepInput = mount(<Stepper type="base" />)
      expect(BaseStepInput.find('.kd-inputNumber-baseStep')).toHaveLength(2)
      const EmbedStepInput = mount(<Stepper type="embed" />)
      expect(EmbedStepInput.find('.kd-inputNumber-embedStep')).toHaveLength(1)
    })

    it('baseStep handleStepChang step', () => {
      let value = '0'
      const onChange = jest.fn((e) => {
        value = e.target.value
      })
      const wrapper = mount(<Stepper type="base" onChange={onChange} value={10} step={10} />)
      wrapper.find('.kd-inputNumber-baseStep').at(0).simulate('mousedown')
      expect(value).toBe('0')
      wrapper.find('.kd-inputNumber-baseStep').at(1).simulate('mousedown')
      expect(value).toBe('20')
    })

    it('embedStep handleStepChang', () => {
      let value = '0'
      const onChange = jest.fn((e) => {
        value = e.target.value
      })
      const wrapper = mount(<Stepper type="embed" onChange={onChange} />)
      wrapper.find('.kd-inputNumber-embedStep-minus').simulate('mousedown')
      expect(value).toBe('-1')
      wrapper.find('.kd-inputNumber-embedStep-plus').simulate('mousedown')
      expect(value).toBe('0')
    })

    it('Stepper renders decimalLength', () => {
      const wrapper = mount(<Stepper placeholder="基本使用" type="base" decimalLength={2} defaultValue={'2.3333'} />)
      const input = wrapper.find('input')
      expect(input.props().value).toEqual('2.33')
    })

    it('Stepper renders  digitLength onChange', async () => {
      let value = ''
      const onChange = jest.fn((e) => {
        value = e.target.value
        expect(value).toBe('2333')
      })
      const wrapper = mount(<Stepper digitLength={4} onChange={onChange} />)
      wrapper.simulate('change', { target: { value: '23332.3333' } })
      // expect(onChange).toHaveBeenCalled()
    })

    it('Stepper renders stepBtnClass', () => {
      const wrapper = mount(<Stepper stepBtnClassName="stepBtnClass" />)
      expect(wrapper.find('.kd-inputNumber-baseStep').first()).toHaveClassName('stepBtnClass')
    })

    it('Stepper renders  symbol', () => {
      let value = ''
      const onChange = jest.fn((e) => {
        value = e.target.value
        expect(value).toBe('$1111')
      })
      const wrapper = mount(<Stepper onChange={onChange} symbol={'$'} />)
      wrapper.simulate('change', { target: { value: '1111' } })
    })

    it('Stepper renders zeroShow', () => {
      let value = ''
      const onChange = jest.fn((e) => {
        value = e.target.value
        expect(value).toBe('')
      })
      const wrapper = mount(<Stepper onChange={onChange} zeroShow={false} />)
      wrapper.simulate('change', { target: { value: '0' } })
    })

    it('defaultValue', () => {
      const wrapper = mount(<Stepper placeholder="基本使用" type="base" defaultValue={'123'} />)
      const input = wrapper.find('input')
      expect(input.prop('value')).toEqual('123')
    })

    it('value', () => {
      const wrapper = mount(<Stepper placeholder="基本使用" type="base" value={'123'} />)
      const input = wrapper.find('input')
      expect(input.prop('value')).toEqual('123')
    })

    it('value defaultValue onChange', () => {
      let value = ''
      const onChange = jest.fn((e) => {
        value = e.target.value
      })
      const wrapper = mount(<Stepper defaultValue={'123'} value={'1234'} onChange={onChange} />)
      const input = wrapper.find('input')
      expect(input.prop('value')).toEqual('1234')
      wrapper.find('.kd-inputNumber-baseStep').at(0).simulate('mousedown')
      expect(value).toEqual('1233')
      expect(input.prop('value')).toEqual('1234')
    })

    it('value onChange', () => {
      let value = ''
      const onChange = jest.fn((e) => {
        value = e.target.value
        wrapper.setProps({ value })
      })
      const wrapper = mount(<Stepper value={'1234'} onChange={onChange} />)
      const input = wrapper.find('input')
      wrapper.find('.kd-inputNumber-baseStep').at(0).simulate('mousedown')
      expect(value).toEqual('1233')
      expect(input.prop('value')).toEqual('1234')
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const Config = {
        compDefaultProps: {
          Stepper: {
            size: 'middle',
          },
        },
      }
      let value = '0'
      const onChange = jest.fn((e) => {
        value = e.target.value
      })
      const wrapper = mount(
        <ConfigProvider value={Config}>
          <Stepper type="base" onChange={onChange} value={10} step={10} />
        </ConfigProvider>,
      )
      wrapper.find('.kd-inputNumber-baseStep').at(0).simulate('mousedown')
      expect(value).toBe('0')
      wrapper.find('.kd-inputNumber-baseStep').at(1).simulate('mousedown')
      expect(value).toBe('20')
    })
  })

  // 9. ref test
  it('should get stepper element from ref', () => {
    const ref = React.createRef()
    mount(<Stepper ref={ref} />)
    expect(ref.current instanceof HTMLInputElement).toBe(true)
  })
})
