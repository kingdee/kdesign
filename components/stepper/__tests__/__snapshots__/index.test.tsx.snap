// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper renders correctly 1`] = `
<span
  class="kd-input-wrapper kd-input-wrapper-size-middle kd-inputNumber kd-inputNumber-stepper"
>
  <span
    class="kd-input-prefix"
  >
    <span
      class="kd-inputNumber-baseStep kd-inputNumber-baseStep-decrease kd-inputNumber-baseStep-middle"
    >
      <i
        class="kdicon kd-inputNumber-icon kdicon-reduce"
      />
    </span>
  </span>
  <input
    class="kd-input kd-input-size-middle"
    prefix="[object Object]"
    stepperrref="[object Object]"
    type="text"
    value=""
  />
  <span
    class="kd-input-suffix"
  >
    <span>
      <span
        class="kd-inputNumber-baseStep kd-inputNumber-baseStep-increase kd-inputNumber-baseStep-middle"
      >
        <i
          class="kdicon kd-inputNumber-icon kdicon-add"
        />
      </span>
    </span>
  </span>
</span>
`;

exports[`Stepper renders correctly 2`] = `
<span
  class="kd-input-wrapper kd-input-wrapper-size-middle kd-inputNumber kd-inputNumber-stepper"
>
  <input
    class="kd-input kd-input-size-middle"
    stepperrref="[object Object]"
    type="text"
    value=""
  />
  <span
    class="kd-input-suffix"
  >
    <span>
      <span
        class="kd-inputNumber-embedStep kd-inputNumber-embedStep-middle"
      >
        <span
          class="kd-inputNumber-embedStep-plus"
        >
          <i
            class="kdicon kd-inputNumber-icon kdicon-arrow-up"
          />
        </span>
        <span
          class="kd-inputNumber-embedStep-minus"
        >
          <i
            class="kdicon kd-inputNumber-icon kdicon-arrow-down"
          />
        </span>
      </span>
    </span>
  </span>
</span>
`;

exports[`Stepper renders correctly 3`] = `
<span
  class="kd-input-wrapper kd-input-wrapper-size-middle kd-inputNumber kd-inputNumber-stepper"
>
  <span
    class="kd-input-prefix"
  >
    <span
      class="kd-inputNumber-baseStep kd-inputNumber-baseStep-decrease kd-inputNumber-baseStep-middle"
    >
      <i
        class="kdicon kd-inputNumber-icon kdicon-reduce"
      />
    </span>
  </span>
  <input
    class="kd-input kd-input-size-middle"
    prefix="[object Object]"
    stepperrref="[object Object]"
    type="text"
    value=""
  />
  <span
    class="kd-input-suffix"
  >
    <span>
      <span
        class="kd-inputNumber-baseStep kd-inputNumber-baseStep-increase kd-inputNumber-baseStep-middle"
      >
        <i
          class="kdicon kd-inputNumber-icon kdicon-add"
        />
      </span>
    </span>
  </span>
</span>
`;
